updater: make UpdateListIter a proper iterator

DemiMarie opened the pull request:
Python iterators are required to implement an `__iter__()` that returns
self [1].  CPython doesn't check this consistently, but the requirement
is still there, so the existing code is buggy.  Python 3.13 started
checking this in list comprehensions, resulting in exceptions being
thrown.  Fix the bug by having `__iter__()` return self, as required by
the iterator protocol.

This worked on Python 3.13 and below, but broke in 3.13.1 [2].

[1]: https://docs.python.org/3/glossary.html#term-iterator
[2]: https://github.com/python/cpython/issues/128211

marmarek commented:
If that's a CPython bug, please report it properly (and then put bug reference in the commit message). But it's more likely buggy usage somewhere (possibly in this repo). Do you have the call trace of problematic situation?

DemiMarie commented:
> If that's a CPython bug, please report it properly (and then put bug reference in the commit message).

I would, if I was able to reproduce it.  The bug went away when I reinstalled, so it seemed like something weird (stale `.pyc` file?).

> But it's more likely buggy usage somewhere (possibly in this repo). Do you have the call trace of problematic situation?

I don’t have the full stack trace, but I do remember that the error happened here:

https://github.com/QubesOS/qubes-desktop-linux-manager/blob/a3e4d177785bbc1b1ee72bd7cd83ef7b8a4e8827/qui/updater/intro_page.py#L166

Python complained that an `UpdatesListIter` was not iterable.  However, the following worked:

```python
    selected_num = 0
    for row in self.list_store:
        selected_num += i.selected
```

Furthermore, I found that `self.list_store` was a `ListWrapper`, not an `UpdatesListIter`.

marmarek commented:
openQA reproduced the issue:
```
Dec 28 17:01:24.186710 dom0 widget-wrapper[9380]: Traceback (most recent call last):
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:   File "/usr/lib/python3.13/site-packages/qui/updater/utils.py", line 48, in wrapper
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:     func(self, *args, **kwargs)
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:     ~~~~^^^^^^^^^^^^^^^^^^^^^^^
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:   File "/usr/lib/python3.13/site-packages/qui/updater/intro_page.py", line 166, in on_checkbox_toggled
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:     selected_num = sum(row.selected for row in self.list_store)
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:   File "/usr/lib/python3.13/site-packages/qui/updater/intro_page.py", line 166, in <genexpr>
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:     selected_num = sum(row.selected for row in self.list_store)
Dec 28 17:01:24.188175 dom0 widget-wrapper[9380]:                                                ^^^^^^^^^^^^^^^
Dec 28 17:01:24.188906 dom0 widget-wrapper[9380]: TypeError: 'UpdateListIter' object is not iterable
Dec 28 17:01:27.597130 dom0 widget-wrapper[9380]: Traceback (most recent call last):
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:   File "/usr/lib/python3.13/site-packages/qui/updater/updater.py", line 215, in next_clicked
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:     vms_to_update = self.intro_page.get_vms_to_update()
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:   File "/usr/lib/python3.13/site-packages/qui/updater/intro_page.py", line 147, in get_vms_to_update
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:     return self.list_store.get_selected()
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:   File "/usr/lib/python3.13/site-packages/qui/updater/utils.py", line 306, in get_selected
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:     selected_rows = [row for row in self if row.selected]
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]:                                     ^^^^
Dec 28 17:01:27.597815 dom0 widget-wrapper[9380]: TypeError: 'UpdateListIter' object is not iterable
```
https://openqa.qubes-os.org/tests/123911#investigation says the python3 version in dom0 didn't change since last good run, but the desktop-linux-manager version did change, so it's most likely a regression in this repo - maybe in the other PR here?

DemiMarie commented:
> https://openqa.qubes-os.org/tests/123911#investigation says the python3 version in dom0 didn't change since last good run, but the desktop-linux-manager version did change, so it's most likely a regression in this repo - maybe in the other PR here?

Possibly.  From what I can tell, the existing code relies on undefined behavior, so it could break at any time.

Did your merge scripts start including the PR message?  In the past I have let that get stale: for instance, this PR is not a workaround for a CPython bug, but rather a bug fix for relying on undocumented behavior in CPython.  Also, the PR message is meant to be viewed as markdown, rather than as plain text, so it has very long lines.

marmarek reviewed (changes requested):
Ugh, in the meantime black formatting is enforced and it doesn't like this: https://gitlab.com/QubesOS/qubes-desktop-linux-manager/-/jobs/8763230803
